qemu/tests/vsock_test.py

```diff
@@ -0,0 +1,152 @@
+    :param port: the port to listen
+    :param session: guest shell session
+    :return: the shell session with opened vsock listening process
+    """
+    nc_vsock_cmd = "%s -l %s" % (nc_vsock_bin, port)
+    session.read_nonblocking(0, timeout=10)
+    logging.info("Listening to the vsock port from guest: %s" % nc_vsock_cmd)
+    session.sendline(nc_vsock_cmd)
+
+
+def nc_vsock_connect(nc_vsock_bin, guest_cid, port):
+    """
+    Connect to vsock port from host, cmd: nc-vsock $guest_cid $port
+
+    :param nc_vsock_bin: path of binary nc-vsock
+    :param guest_cid: guest cid to connect
+    :param port: port to connect
+    :return: The vsock session from host side, being waiting for input
+    """
+    nc_vsock_cmd = "%s %s %s" % (nc_vsock_bin, guest_cid, port)
+    logging.info("Connect to the vsock port on host: %s" % nc_vsock_cmd)
+    return aexpect.Expect(
+        nc_vsock_cmd,
+        auto_close=False,
+        output_func=utils_misc.log_line,
+        output_params=("vsock_%s_%s" % (guest_cid, port),))
+
+
+@error_context.context_aware
+def run(test, params, env):
+    """
+    Vsock basic function test
+
+    1. Boot guest with vhost-vsock-pci device
+    2. Download and compile nc-vsock on both guest and host
+    3. Start listening inside guest, nc-vsock -l $port
+    4. Connect guest CID from host, nc-vsock $guest_cid $port
+    5. Input character, e.g. 'Hello world'
+    6. Check if guest receive the content correctly
+
+    :param test: QEMU test object
+    :param params: Dictionary with the test parameters
+    :param env: Dictionary with test environment
+    """
+
+    def clean(tmp_file=None):
+        """ Clean the environment """
+        cmd_rm = 'rm -rf %s*' % nc_vsock_bin
+        if tmp_file:
+            cmd_rm += '; rm -rf %s' % tmp_file
+        session.cmd_output_safe(cmd_rm)
+        process.system(cmd_rm, ignore_status=True)
+        if host_vsock_session.is_alive():
+            host_vsock_session.close()
+        session.close()
+
+    vm = env.get_vm(params["main_vm"])
+    tmp_file = '/tmp/vsock_file_%s' % utils_misc.generate_random_string(6)
+    session = vm.wait_for_login()
+    nc_vsock_bin = compile_nc_vsock(test, vm, session)
+    vm_ip_addr = vm.wait_for_get_address(0)
+    port = utils_misc.find_free_port(5000, 6000, vm_ip_addr)
+    vsock_dev = params["vsocks"].split()[0]
+    guest_cid = vm.devices.get(vsock_dev).get_param("guest-cid")
+    nc_vsock_listen(nc_vsock_bin, port, session)
+    host_vsock_session = nc_vsock_connect(nc_vsock_bin, guest_cid, port)
+    connected_str = r"Connection from cid*"
+    send_data = "Hello world"
+    if not session.read_until_last_line_matches(connected_str):
+        clean()
+        test.fail("Connection failed.")
+    error_context.context('Input "Hello world" to vsock.', logging.info)
+    host_vsock_session.sendline(send_data)
+    if not session.read_until_last_line_matches(send_data):
+        clean()
+        test.fail("Data transfer failed.")
```

Comment:
The way to check `send_data` should have some problem.
As the selinux problem, this step should fail too (test manually, it failed), but this step pass, case only failed when compare md5value.
I guess,  `session` (get from wait_for_log) is not the expect one, use guest's nc-vsock listening session ?  Not very clear about this.

Reply:
@yanan-fu I am not sure your issue with manual test, but I do get it failed when the connection not being established correctly:
```
2019-02-25 22:16:32,291 client           L1162 DEBUG| Sending command (safe): /home/vsock_test/nc-vsock -l 5000 > /root/vsock_3_5000 &
2019-02-25 22:16:32,394 vsock_test       L0062 INFO | Connect to the vsock port on host: /home/vsock_test/nc-vsock 3 5000
2019-02-25 22:16:32,456 utils_misc       L0586 DEBUG| Wait connection from host (0.000023 secs)
2019-02-25 22:16:32,457 client           L1126 DEBUG| Sending command: cat /root/vsock_3_5000
2019-02-25 22:16:32,562 utils_misc       L0586 DEBUG| Wait receive data from host (0.000007 secs)
2019-02-25 22:16:32,562 client           L1126 DEBUG| Sending command: cat /root/vsock_3_5000
2019-02-25 22:16:34,668 utils_misc       L0586 DEBUG| Wait receive data from host (2.106245 secs)
2019-02-25 22:16:34,669 client           L1126 DEBUG| Sending command: cat /root/vsock_3_5000
2019-02-25 22:16:36,775 utils_misc       L0586 DEBUG| Wait receive data from host (4.213789 secs)
2019-02-25 22:16:36,776 client           L1126 DEBUG| Sending command: cat /root/vsock_3_5000
2019-02-25 22:16:38,884 utils_misc       L0586 DEBUG| Wait receive data from host (6.322144 secs)
2019-02-25 22:16:38,884 client           L1126 DEBUG| Sending command: cat /root/vsock_3_5000
2019-02-25 22:16:40,991 utils_misc       L0586 DEBUG| Wait receive data from host (8.429545 secs)
2019-02-25 22:16:40,992 client           L1126 DEBUG| Sending command: cat /root/vsock_3_5000
2019-02-25 22:16:43,274 stacktrace       L0041 ERROR|
2019-02-25 22:16:43,275 stacktrace       L0044 ERROR| Reproduced traceback from: /usr/lib/python2.7/site-packages/avocado_plugins_vt-68.0-py2.7.egg/avocado_vt/test.py:438
2019-02-25 22:16:43,276 stacktrace       L0047 ERROR| Traceback (most recent call last):
2019-02-25 22:16:43,276 stacktrace       L0047 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_plugins_vt-68.0-py2.7.egg/virttest/error_context.py", line 135, in new_fn
2019-02-25 22:16:43,276 stacktrace       L0047 ERROR|     return fn(*args, **kwargs)
2019-02-25 22:16:43,276 stacktrace       L0047 ERROR|   File "/var/lib/avocado/data/avocado-vt/test-providers.d/downloads/io-github-autotest-qemu/qemu/tests/vsock_test.py", line 121, in run
2019-02-25 22:16:43,277 stacktrace       L0047 ERROR|     test.fail("Data transfer failed.")
2019-02-25 22:16:43,277 stacktrace       L0047 ERROR|   File "/usr/lib/python2.7/site-packages/avocado_framework-67.0-py2.7.egg/avocado/core/test.py", line 1025, in fail
2019-02-25 22:16:43,277 stacktrace       L0047 ERROR|     raise exceptions.TestFail(message)
2019-02-25 22:16:43,277 stacktrace       L0047 ERROR| TestFail: Data transfer failed.
```
Guset session is listening instead of connecting or writing, no really interaction in the session, so it  is meaningless to start individual Except process for it.

Reply:
@vivianQizhu AS we discussed,  the result above is not the latest code.
Test with the latest code, debug log show:
```
03:30:48 INFO | Listening to the vsock port from guest: /home/nc-vsock -l 5000
03:30:48 INFO | Connect to the vsock port on host: /home/nc-vsock 3 5000
03:30:48 INFO | Context: Input "Hello world" to vsock.
03:30:48 DEBUG| Attempting to log into 'avocado-vt-vm1' (timeout 240s)
03:30:48 DEBUG| Found/Verified IP 10.73.73.191 for VM avocado-vt-vm1 NIC 0
03:30:50 DEBUG| Sending command: dd if=/dev/urandom of=/tmp/vsock_file_a5XS0F count=1000 bs=1M
```
It failed when compare md5 value, while send `Hello word` should failed to as the selinux problem.

Reply:
@yanan-fu Seems the product issue only exist on 8, I did not test on it yet, 7 pass all the test, so let me check it on 8 first. Thanks.